src/python/dxpy/bindings/download_all_inputs.py

```diff
@@ -0,0 +1,161 @@
+# Copyright (C) 2014-2015 DNAnexus, Inc.
+#
+# This file is part of dx-toolkit (DNAnexus platform client libraries).
+#
+#   Licensed under the Apache License, Version 2.0 (the "License"); you may not
+#   use this file except in compliance with the License. You may obtain a copy
+#   of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#   License for the specific language governing permissions and limitations
+#   under the License.
+
+import concurrent.futures
+import os
+import sys
+import multiprocessing
+from collections import defaultdict
+
+import dxpy
+from dxpy.utils import file_load_utils
+
+def _create_dirs(idir, dirs):
+    '''
+    Create a set of directories, so we could store the input files.
+    For example, seq1 could be stored under:
+        /in/seq1/NC_001122.fasta
+
+    TODO: this call could fail, we need to report a reasonable error code
+
+    Note that we create a directory for every file array, even if
+    it has zero inputs.
+    '''
+    # create the <idir> itself
+    file_load_utils.ensure_dir(idir)
+    # create each subdir
+    for d in dirs:
+        file_load_utils.ensure_dir(os.path.join(idir, d))
+
+def _download_one_file(file_rec, idir):
+    src_file = file_rec['src_file_id']
+    trg_file = os.path.join(idir, file_rec['trg_fname'])
+    print("downloading file: " + src_file + " to filesystem: " + trg_file)
+    dxpy.download_dxfile(src_file, trg_file)
+    return file_rec
+
+# Download the files sequentially
+#   to_download: list of tuples describing files to download
+def _sequential_file_download(to_download, idir):
+    for file_rec in to_download:
+        _download_one_file(file_rec, idir)
+
+# Download files in parallel
+#   to_download: list of tuples describing files to download
+def _parallel_file_download(to_download, idir, max_num_parallel_downloads):
+    with concurrent.futures.ThreadPoolExecutor(max_workers=max_num_parallel_downloads) as executor:
+       future_files = {executor.submit(_download_one_file, file_rec, idir): file_rec for file_rec in to_download}
+       for future in concurrent.futures.as_completed(future_files):
+           file_rec = future_files[future]
+           try:
+               future.result()
+           except Exception as exc:
```

Comment:
I have updated the PR to include a try/except block wrapping the concurrent.futures block with the same pattern as the example @kislyuk pointed to.  I also added the suggested timeout.